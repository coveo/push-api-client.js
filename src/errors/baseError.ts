export abstract class PushApiClientBaseError extends Error {
  public name = 'Push API Client Error';
  public constructor(message?: string) {
    super(message);
  }
}
