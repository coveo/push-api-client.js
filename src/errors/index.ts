export * from './fieldErrors';
export * from './privilegeError';
export * from './validatorErrors';
