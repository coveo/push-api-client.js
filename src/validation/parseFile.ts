import dedent from 'ts-dedent';
import {CompressionType, CompressionTypeEnum} from './../document';
import {PrimitivesValues, StringValue, EnumValue} from '@coveo/bueno';
import {DocumentBuilder, MetadataValue} from '..';
import {existsSync, lstatSync, PathLike, readFileSync} from 'fs';
import {CaseInsensitiveDocument} from './caseInsensitiveDocument';
import {KnownKeys} from './knownKey';
import {
  InvalidDocument,
  NotAFileError,
  NotAJsonFileError,
} from '../errors/validatorErrors';
import {RequiredKeyValidator} from './requiredKeyValidator';
import {Metadata} from '../document';
import {ParseDocumentOptions} from '../interfaces';
import {processPermissionList} from './parsePermissions';

export const parseAndGetDocumentBuilderFromJSONDocument = async (
  documentPath: PathLike,
  options?: ParseDocumentOptions
): Promise<DocumentBuilder[]> => {
  ensureFileIntegrity(documentPath);

  const fileContent = safeJSONParse(documentPath);

  const executeCallback = async (docBuilder: DocumentBuilder) => {
    if (options?.callback) {
      await options?.callback(docBuilder, documentPath);
    }
  };

  if (Array.isArray(fileContent)) {
    const docBuilders: DocumentBuilder[] = [];
    for (const doc of fileContent) {
      const docBuilder = processDocument(doc, documentPath, options);
      await executeCallback(docBuilder);
      docBuilders.push(docBuilder);
    }
    return docBuilders;
  } else {
    const docBuilder = processDocument(fileContent, documentPath, options);
    executeCallback(docBuilder);
    return [docBuilder];
  }
};

const safeJSONParse = (documentPath: PathLike) => {
  const fileContent = readFileSync(documentPath).toString();
  try {
    return JSON.parse(fileContent) as
      | Record<string, string>
      | Record<string, string>[];
  } catch (error) {
    throw new NotAJsonFileError(documentPath);
  }
};

const processDocument = (
  fileContent: Record<string, PrimitivesValues>,
  documentPath: PathLike,
  options?: ParseDocumentOptions
) => {
  const caseInsensitiveDoc = new CaseInsensitiveDocument(fileContent);

  const documentBuilder = validateRequiredKeysAndGetDocumentBuilder(
    caseInsensitiveDoc,
    documentPath
  );
  try {
    processKnownKeys(caseInsensitiveDoc, documentBuilder);
    processPermissionList(caseInsensitiveDoc, documentBuilder, documentPath);
    processMetadata(caseInsensitiveDoc, documentBuilder, options);
  } catch (error) {
    if (typeof error === 'string') {
      throw new InvalidDocument(
        documentPath,
        caseInsensitiveDoc.originalDocument,
        error
      );
    }
    throw error;
  }
  return documentBuilder;
};

const validateRequiredKeysAndGetDocumentBuilder = (
  caseInsensitiveDoc: CaseInsensitiveDocument<PrimitivesValues>,
  documentPath: PathLike
) => {
  const requiredDocumentId = new RequiredKeyValidator<string>(
    ['documentid', 'uri'],
    caseInsensitiveDoc,
    new StringValue({required: true, emptyAllowed: false})
  );

  if (!requiredDocumentId.isValid) {
    throw new InvalidDocument(
      documentPath,
      caseInsensitiveDoc.originalDocument,
      requiredDocumentId.explanation
    );
  }

  const requiredDocumentTitle = new RequiredKeyValidator<string>(
    'title',
    caseInsensitiveDoc,
    new StringValue({required: true, emptyAllowed: false})
  );
  if (!requiredDocumentTitle.isValid) {
    throw new InvalidDocument(
      documentPath,
      caseInsensitiveDoc.originalDocument,
      requiredDocumentTitle.explanation
    );
  }

  caseInsensitiveDoc.remove('documentid', 'uri', 'title');

  return new DocumentBuilder(
    requiredDocumentId.value!,
    requiredDocumentTitle.value!
  );
};

const processKnownKeys = (
  caseInsensitiveDoc: CaseInsensitiveDocument<PrimitivesValues>,
  documentBuilder: DocumentBuilder
) => {
  new KnownKeys<string>('author', caseInsensitiveDoc).whenExists((author) => {
    documentBuilder.withAuthor(author);
    caseInsensitiveDoc.remove('author');
  });
  new KnownKeys<string>('clickableuri', caseInsensitiveDoc).whenExists(
    (clickuri) => {
      documentBuilder.withClickableUri(clickuri);
      caseInsensitiveDoc.remove('clickableuri');
    }
  );
  new KnownKeys<string>('data', caseInsensitiveDoc).whenExists((data) => {
    documentBuilder.withData(data);
    caseInsensitiveDoc.remove('data');
  });
  new KnownKeys<string>('date', caseInsensitiveDoc).whenExists((date) => {
    documentBuilder.withDate(date);
    caseInsensitiveDoc.remove('date');
  });
  new KnownKeys<string>('modifieddate', caseInsensitiveDoc).whenExists(
    (modifiedDate) => {
      documentBuilder.withModifiedDate(modifiedDate);
      caseInsensitiveDoc.remove('modifieddate');
    }
  );
  new KnownKeys<string>('fileextension', caseInsensitiveDoc).whenExists(
    (fileExtension) => {
      documentBuilder.withFileExtension(fileExtension);
      caseInsensitiveDoc.remove('fileextension');
    }
  );
  new KnownKeys<string>('permanentid', caseInsensitiveDoc).whenExists(
    (permanentId) => {
      documentBuilder.withPermanentId(permanentId);
      caseInsensitiveDoc.remove('permanentid');
    }
  );
  new KnownKeys<string>('compressedBinaryData', caseInsensitiveDoc).whenExists(
    (compressedBinaryData) =>
      processCompressedBinaryData(
        compressedBinaryData,
        caseInsensitiveDoc,
        documentBuilder
      )
  );
};

const processCompressedBinaryData = (
  compressedBinaryData: string,
  caseInsensitiveDoc: CaseInsensitiveDocument<PrimitivesValues>,
  documentBuilder: DocumentBuilder
) => {
  const compressionType = new RequiredKeyValidator<CompressionType>(
    ['compressionType'],
    caseInsensitiveDoc,
    new EnumValue({enum: CompressionTypeEnum, required: false})
  );

  if (!compressionType.isValid || compressionType.value === null) {
    const options: CompressionTypeEnum[] = [];
    for (const compressionType in CompressionTypeEnum) {
      options.push(
        CompressionTypeEnum[compressionType as keyof typeof CompressionTypeEnum]
      );
    }
    throw dedent`${compressionType.explanation}
    Supported values are: ${options.join(', ')}`;
  }

  documentBuilder.withCompressedBinaryData(
    compressedBinaryData,
    compressionType.value
  );
  caseInsensitiveDoc.remove('compressedBinaryData');
  caseInsensitiveDoc.remove('compressionType');
};

const processMetadata = (
  caseInsensitiveDoc: CaseInsensitiveDocument<PrimitivesValues>,
  documentBuilder: DocumentBuilder,
  options?: ParseDocumentOptions
) => {
  const metadata: Metadata = {};
  Object.entries(caseInsensitiveDoc.remainingRecord).forEach(([k, v]) => {
    metadata[k] = v! as Extract<PrimitivesValues, MetadataValue>;
  });
  documentBuilder.withMetadata(metadata, options?.fieldNameTransformer);
};

const ensureFileIntegrity = (documentPath: PathLike) => {
  if (!isFile(documentPath)) {
    throw new NotAFileError(documentPath);
  }
};

const isFile = (p: PathLike) => {
  if (!existsSync(p)) {
    return false;
  }
  return lstatSync(p).isFile();
};
