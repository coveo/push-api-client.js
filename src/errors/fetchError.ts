import {PushApiClientBaseError} from './baseError';
import dedent from 'ts-dedent';

export class FetchError extends PushApiClientBaseError {
  public name = 'Fetch Error';
  public static async build(response: Response) {
    let body: string;
    try {
      body = JSON.stringify(await response.json());
    } catch (error) {
      body = await response.text();
    }
    return new FetchError(body);
  }
  private constructor(public body: string) {
    super();
    this.message = dedent`
      Response body:
      ${body}
    `;
  }
}
