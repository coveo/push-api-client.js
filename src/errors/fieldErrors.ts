import {FieldModel} from '@coveord/platform-client';
import {Inconsistencies} from '../fieldAnalyser/inconsistencies';
import {PushApiClientBaseError} from './baseError';

export class FieldTypeInconsistencyError extends PushApiClientBaseError {
  public name = 'Field Type Inconsistency Error';
  public constructor(inconsitencies: Inconsistencies) {
    super();
    this.message = `Inconsistent types detected in the following field${
      inconsitencies.size > 1 ? 's' : ''
    }:`;
    inconsitencies.forEach((typeSet, field) => {
      const inconsistentTypes = [];
      for (const type of typeSet) {
        inconsistentTypes.push(type);
      }
      this.message += `
      ${field}: ${inconsistentTypes.sort().join(', ')}`;
    });

    // TODO: CDX-844: display document with type inconsistencies
    this.message += `
      Make sure to review and fix your document metadata before pushing`;
  }
}

export class InvalidPermanentId extends PushApiClientBaseError {
  public name = 'Invalid permanentId field Error';
  public constructor(field: FieldModel) {
    super(`
The permanentid field detected in the index is not correctly configured.
Expected field type: STRING
Current field type: ${field.type}

You can delete and recreate the permanentid field as a STRING field.
See docs.coveo.com/en/1913 fore more info.
    `);
  }
}
export class UnsupportedFieldError extends PushApiClientBaseError {
  public name = 'Unsupported field Error';
  public constructor(public readonly unsupportedFields: string[]) {
    super(`
The following field names are invalid:
${unsupportedFields.reduce(
  (prev: string, curr: string) => (prev += `  â€¢ ${curr}\n`),
  ''
)}

Field names can only contain lowercase letters (a-z), numbers (0-9), and underscores. The field name must be at least one character long and must start with a lowercase letter.
    `);
  }
}
