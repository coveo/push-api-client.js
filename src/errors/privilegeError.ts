import {PushApiClientBaseError} from './baseError';

export class PrivilegeError extends PushApiClientBaseError {
  public name = 'Privilege Error';
  public constructor(public message: string) {
    super(message);
  }
}
