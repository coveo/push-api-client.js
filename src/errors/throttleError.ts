import {PushApiClientBaseError} from './baseError';

export class ThrottleError extends PushApiClientBaseError {
  public name = 'Throttle Error';
  public constructor() {
    super();
  }
}
