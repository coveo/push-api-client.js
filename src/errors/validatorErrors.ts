import {PrimitivesValues} from '@coveo/bueno';
import {PathLike} from 'fs';
import {PushApiClientBaseError} from './baseError';

export class NotAFileError extends PushApiClientBaseError {
  public name = 'Not A File Error';
  public constructor(p: PathLike) {
    super(`${p} is not a valid file, or does not exists.`);
  }
}
export class NotAJsonFileError extends PushApiClientBaseError {
  public name = 'Not A JSON File Error';
  public constructor(p: PathLike) {
    super(`${p} is not a valid JSON file.`);
  }
}

export class InvalidDocument extends PushApiClientBaseError {
  public name = 'Invalid JSON Document Error';
  private doc: Readonly<Record<string, PrimitivesValues>>;
  public constructor(
    p: PathLike,
    doc: Record<string, PrimitivesValues>,
    explanation: string
  ) {
    super(
      [
        `${p} is not a valid JSON: ${explanation}`,
        'Helpful links on the expected JSON format:',
        ' • JSON file example: https://github.com/coveo/push-api-client.js/tree/main/samples/json',
        ' • Document Body reference: https://docs.coveo.com/en/75#documentbody',
      ].join('\n')
    );
    this.doc = doc;
  }

  public get document() {
    return this.doc;
  }
}

export class UnsupportedAttribute extends PushApiClientBaseError {
  public name = 'Unsupported Attribute Error';
  public constructor(p: PathLike, unsupported: string) {
    super(
      `${p} contains a currently unsupported document attribute: ${unsupported}`
    );
  }
}
