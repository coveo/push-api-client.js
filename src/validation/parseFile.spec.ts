import {join} from 'path';
import {cwd} from 'process';
import {parseAndGetDocumentBuilderFromJSONDocument} from './parseFile';
import {InvalidDocument} from '../errors/validatorErrors';

describe('parseFile', () => {
  const pathToStub = join(cwd(), 'src', '__stub__');
  const parse = (file: string) => () =>
    parseAndGetDocumentBuilderFromJSONDocument(file);

  it('should accept non-url documentIDs', async () => {
    const file = join(pathToStub, 'jsondocuments', 'notAnUrl.json');
    await expect(
      parseAndGetDocumentBuilderFromJSONDocument(file)
    ).resolves.not.toThrow();
  });

  it.each([
    {
      title: 'should fail on missing title',
      fileName: 'noTitle.json',
      error: 'title: value is required.',
    },
    {
      title: 'should fail on missing documentID',
      fileName: 'noID.json',
      error: 'documentid: value is required.',
    },
    {
      title: 'should fail on missing identity',
      fileName: 'noIdentity.json',
      error: 'allowedpermissions:  value does not contain identity',
    },
    {
      title: 'should fail on identity with an invalid string',
      fileName: 'identityNotAString.json',
      error: 'allowedpermissions:   value is not a string.',
    },
    {
      title: 'should fail on allow anonymous with an invalid boolean',
      fileName: 'identityAllowAnonymousNotABoolean.json',
      error: 'allowanonymous: value is not a boolean.',
    },
    {
      title: 'should fail on identityType with an invalid value',
      fileName: 'identityTypeInvalidValue.json',
      error:
        'allowedpermissions:   value should be one of: UNKNOWN, USER, GROUP, VIRTUAL_GROUP.',
    },
  ])('$title', async ({fileName, error}) => {
    const file = join(pathToStub, 'jsondocuments', fileName);
    await expect(parse(file)).rejects.toThrow(
      new InvalidDocument(
        file,
        `Document contains an invalid value for ${error}`
      )
    );
  });

  it('should fail on reserved keyword', async () => {
    const file = join(pathToStub, 'jsondocuments', 'reservedKeyword.json');
    await expect(parse(file)).rejects.toThrow(
      new InvalidDocument(
        file,
        'Cannot use parentid as a metadata key: It is a reserved key name. See https://docs.coveo.com/en/78/index-content/push-api-reference#json-document-reserved-key-names'
      )
    );
  });

  it('should fail on unsupported metadata key', async () => {
    const file = join(pathToStub, 'jsondocuments', 'invalidFields.json');
    await expect(parse(file)).rejects.toThrowErrorMatchingSnapshot();
  });

  it.todo('should only push JSON documents'); // TODO: from mix documents
});
