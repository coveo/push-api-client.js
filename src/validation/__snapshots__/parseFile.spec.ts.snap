// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseFile should fail on unsupported metadata key 1`] = `
"
The following field names are still invalid after transformation:
  • <empty field name> --> <empty field name>
  • 0invalid-field --> 0invalid-field
  • poster-a --> poster-a

Field names can only contain lowercase letters (a-z), numbers (0-9), and underscores. Field names must be at least one character long and must start with a lowercase letter.
    "
`;
